### PR TITLE
feat: add project slugs to organization endpoint

### PR DESCRIPTION
**Pull request by nhsiehgit**
Enables fetching projects for an organization by slugs rather than by project ids

included a catchall constant `$all` which is an invalid slug name and much easier to understand than `-1`

https://getsentry.atlassian.net/browse/ER-987

### CODE REVIEW

File: src/sentry/constants.py [hunk 1]
```diff
@@ -549,6 +549,7 @@ def from_str(cls, string: str) -> Optional[int]:
 INTERNAL_INTEGRATION_TOKEN_COUNT_MAX = 20
 
 ALL_ACCESS_PROJECTS = {-1}
+ALL_ACCESS_PROJECTS_SLUG = {"$all"}
```
Comment (dashed) [inline src/sentry/constants.py:552]: @nhsiehgit Would love a test for this. 

Is `$all` a project slug that's impossible to set?

Comment (nhsiehgit) [reply src/sentry/constants.py:552]: Yessir - sorry i have a habit of getting the code out first, then working on tests next 😬 
working on that now.

@dashed - yes, slugs starting in `$` are invalid. (as well as other special chars i believe)

File: src/sentry/api/bases/organization.py [hunk 2]
```diff
@@ -195,6 +195,23 @@ def get_projects(
...
+        slugs = request.GET.getlist("projectSlug")
+        if slugs == ALL_ACCESS_PROJECTS_SLUG:
+            p_ids = ALL_ACCESS_PROJECTS
+        else:
+            projects = Project.objects.filter(
+                organization=organization, slug__in=slugs
+            ).values_list("id", flat=True)
+            p_ids = set(projects)
+
+        return self.get_projects(
```
Comment (nhsiehgit) [inline src/sentry/api/bases/organization.py:208]: Note - `get_projects` is doing a lot of project validations as well which is why we're simply passing through to this method.

**Review by wedamija — COMMENTED** [review]
Looks good to me. What's the plan for this? Are we intending to convert existing apis to use it?

File: src/sentry/constants.py (hunk 1 above)
Comment (JoshFerge) [reply src/sentry/constants.py:552]: @nhsiehgit its probably worth adding it to `RESERVED_PROJECT_SLUGS` just in case in 10 years we change how slugs are validated or something :P

**Comment by nhsiehgit** [comment]
> Are we intending to convert existing apis to use it?

@wedamija 
No plans for this currently.
I was chatting with @JoshFerge on some of our quota api's and this came out of some work i had been doing to dedup some api calls.

TLDR: we have an api querying by project id, and have a second api filtering by slug. 
So the solutions are: 
- do nothing and just remember to update both places if we ever make changes to the apis, 
- Only use the id api and make a separate request to fetch the full project via the slug, then make a request with the project id
- Modify our FE routes to pass project id rather than slugs, so we don't need to fetch the full project obj for just the id
- Modify the org projects fetch so we can pass slugs

The latter here seems the most proper and also useful.

re: 
https://sentry.slack.com/archives/C1B4LB39D/p1647984705733689

Sounds like this is a piece of debt we've never really had a need/desire to change before?

**Comment by wedamija** [comment]
> > Are we intending to convert existing apis to use it?
> 
> @wedamija No plans for this currently. I was chatting with @JoshFerge on some of our quota api's and this came out of some work i had been doing to dedup some api calls.
> 
> TLDR: we have an api querying by project id, and have a second api filtering by slug. So the solutions are:
> 
> * do nothing and just remember to update both places if we ever make changes to the apis,
> * Only use the id api and make a separate request to fetch the full project via the slug, then make a request with the project id
> * Modify our FE routes to pass project id rather than slugs, so we don't need to fetch the full project obj for just the id
> * Modify the org projects fetch so we can pass slugs
> 
> The latter here seems the most proper and also useful.
> 
> re: https://sentry.slack.com/archives/C1B4LB39D/p1647984705733689
> 
> Sounds like this is a piece of debt we've never really had a need/desire to change before? edit: Oh - actually this thread was with you too haha

So is the goal to convert one specific api to accept both project ids and slugs?

**Comment by nhsiehgit** [comment]
> So is the goal to convert one specific api to accept both project ids and slugs?

Currently yes, that's correct

File: src/sentry/api/bases/organization.py [hunk 3]
```diff
@@ -190,9 +193,23 @@ def get_projects(
         :return: A list of Project objects, or raises PermissionDenied.
         """
         if project_ids is None:
-            project_ids = self.get_requested_project_ids_unchecked(request)
+            slugs = project_slugs or set(list(filter(None, request.GET.getlist("projectSlug"))))
```
Comment (JoshFerge) [inline src/sentry/api/bases/organization.py:196]: can this just be `set(request.GET.getlist("projectSlug")`?

Comment (nhsiehgit) [reply src/sentry/api/bases/organization.py:196]: Oops - i should remove the `list` yea lol

Comment (nhsiehgit) [reply src/sentry/api/bases/organization.py]: The filter is necessary since it's still possible to pass an empty string or something in the slug in which case we wouldn't be falling back to the id's

File: tests/sentry/api/bases/test_organization.py [hunk 4]
```diff
@@ -310,6 +311,80 @@ def test_all_accessible_sigil_value_allow_joinleave(self):
         self.create_team_membership(user=self.user, team=self.team_1)
         self.run_test([self.project_1, self.project_2], project_ids=[-1])
 
+    @mock.patch("sentry.api.bases.organization.OrganizationEndpoint._get_projects_by_id")
```
Comment (wedamija) [inline tests/sentry/api/bases/test_organization.py:314]: Just personal opinion but I feel like mocking methods like this is more likely to break in the future if the signature changes.

**Review by wedamija — APPROVED** [review]

File: tests/sentry/api/bases/test_organization.py (hunk 4 above)
Comment (nhsiehgit) [reply tests/sentry/api/bases/test_organization.py:314]: Heh. i'm on the opposite side here.

IMO we should be decoupling our test dependencies and properly testing individual components.
We can still have integration tests, but if i'm trying to test a new feature, i shouldn't have to initialize an entire suite of unrelated dependencies.

IMO this is probably what's causing a lot of our flakes and making those flakes incredibly difficult to track down 🤷

Comment (wedamija) [reply tests/sentry/api/bases/test_organization.py:314]: Yeah, I know the reasoning for doing unit tests like this, just my personal opinion that it produces tests that catch fewer issues in practice. 

fwiw our flakes are usually caused by date related issues in tests that are explicitly making queries. I don't think a lack of mocking is usually the cause here.
